packages/remix-server-runtime/server.ts

```diff
@@ -75,15 +85,19 @@ export const createRequestHandler: CreateRequestHandlerFunction = (
     if (url.searchParams.has("_data")) {
       let routeId = url.searchParams.get("_data")!;
 
-      response = await handleDataRequestRR(
+      response = await handleRemixDataRequest(
         serverMode,
         staticHandler,
         routeId,
         request,
-        loadContext
+        loadContext,
+        unstable_middleware ? middlewareContext : undefined
       );
 
-      if (build.entry.module.handleDataRequest) {
+      if (
+        unstable_middleware !== true &&
+        build.entry.module.handleDataRequest
+      ) {
```

Comment:
Just like we will now ignore `getLoadContext` when you opt into middleware, we will also ignore `handleDataRequest` since you can operate on responses via the `serverMiddleware`.  If folks need to decipher between document and data requests they should be able to do that via looking at `response` headers or the `_data` query param